core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1241,8 +1253,37 @@ private[spark] class BlockManager(
     futureExecutionContext.shutdownNow()
     logInfo("BlockManager stopped")
   }
-}
 
+  /**
+   * Contact all external shuffle services and de-register this application, cleaning
+   * shuffle files as well.
+   *
+   * This is required in Mesos mode with dynamic allocation, since executors leave behind
+   * shuffle files to be served by the external shuffle service. We need to delete those
+   * files when the application stops.
+   */
+  private def cleanupAllShuffleFiles() {
```

Comment:
This is only best-effort. If the driver is killed forcefully we still have shuffle files lingering around.